clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -171,7 +171,25 @@ public class TransactionManager {
      *
      * See KAFKA-14831 for more detail.
      */
-    private final ThreadLocal<Boolean> shouldPoisonStateOnInvalidTransition;
+    public interface InvalidTransitionAttemptPolicy {
```

Comment:
Excuse me, are there more policy in the future?

Reply:
You're right, it's not the best name/approach. This behavior could probably just be a boolean passed in during creation.

Reply:
Actually, I take it back. The value can't be static, the behavior is dependent on which thread it's being executed on 🤷

Reply:
The policy only changes when running tests. In those cases there aren't necessarily different threads running, but the test needs to mock that there are.

Reply:
I've updated the approach to use a method named `shouldPoisonStateOnInvalidTransition()` which uses the default behavior (return `true` only if called from the `Sender` thread) and a subclass in `TransactionManagerTest` can optionally override that behavior, where needed.

Does that seem better?